lib/master.go

```diff
@@ -0,0 +1,351 @@
+	ID() MasterID
+	Close(ctx context.Context) error
+}
+
+type MasterImpl interface {
+	Master
+
+	// InitImpl provides customized logic for the business logic to initialize.
+	InitImpl(ctx context.Context) error
+
+	// Tick is called on a fixed interval.
+	Tick(ctx context.Context) error
+
+	// OnWorkerDispatched is called when a request to launch a worker is finished.
+	OnWorkerDispatched(worker WorkerHandle, result error) error
+
+	// OnWorkerOnline is called when the first heartbeat for a worker is received.
+	OnWorkerOnline(worker WorkerHandle) error
+
+	// OnWorkerOffline is called when a worker exits or has timed out.
+	OnWorkerOffline(worker WorkerHandle, reason error) error
+
+	// OnWorkerMessage is called when a customized message is received.
+	OnWorkerMessage(worker WorkerHandle, topic p2p.Topic, message interface{}) error
+
+	// CloseImpl is called when the master is being closed
+	CloseImpl(ctx context.Context) error
+}
+
+type BaseMaster struct {
+	MasterImpl
+
+	messageHandlerManager p2p.MessageHandlerManager
+	messageRouter         p2p.MessageRouter
+	metaKVClient          metadata.MetaKV
+	executorClientManager *client.Manager
+	serverMasterClient    *client.MasterClient
+	workers               *workerManager
+	pool                  workerpool.AsyncPool
+
+	// read-only fields
+	currentEpoch atomic.Int64
+	id           MasterID
+
+	// TODO put these two fields somewhere else.
+	advertiseAddr string
+	nodeID        p2p.NodeID
+
+	wg    sync.WaitGroup
+	errCh chan error
+}
+
+func NewBaseMaster(
+	impl MasterImpl,
+	id MasterID,
+	messageHandlerManager p2p.MessageHandlerManager,
+	messageRouter p2p.MessageRouter,
+	metaKVClient metadata.MetaKV,
+	executorClientManager *client.Manager,
+	serverMasterClient *client.MasterClient,
+) *BaseMaster {
+	return &BaseMaster{
+		MasterImpl:            impl,
+		messageHandlerManager: messageHandlerManager,
+		messageRouter:         messageRouter,
+		metaKVClient:          metaKVClient,
+		executorClientManager: executorClientManager,
+		serverMasterClient:    serverMasterClient,
+		pool:                  workerpool.NewDefaultAsyncPool(4),
+		id:                    id,
+	}
+}
+
+func (m *BaseMaster) MetaKVClient() metadata.MetaKV {
+	return m.metaKVClient
+}
+
+func (m *BaseMaster) Init(ctx context.Context) error {
+	// TODO think about what ctx to use to call startBackgroundTasks.
+	m.startBackgroundTasks(ctx)
+
+	if err := m.initMessageHandlers(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	isInit, epoch, err := m.initMetadata(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	m.currentEpoch.Store(epoch)
+	m.workers = newWorkerManager(m.id, !isInit, epoch)
+	if err := m.InitImpl(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := m.markInitializedInMetadata(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (m *BaseMaster) Poll(ctx context.Context) error {
+	OfflinedWorkers := m.workers.Tick(ctx, m.messageRouter)
```

Comment:
the workers.Tick() should be called in the init()?,  the Poll will make the heartbeat timeout